### PR TITLE
Skipping bounties

### PR DESCRIPTION
**Pull request by Killerqu00**
## About the PR

Quartermasters can now skip a single bounty once every 15 minutes.

## Why / Balance

It is mostly a QoL change. Some bounties have items that are not renewable, meaning that a certain bounty can become uncompletable.

## Technical details

StationCargoBountyDatabaseComponent contains TimeSpan of next time when bounty skipping is available. CargoSystem.Bounty / OnSkipBountyMessage contains the logic, such as checks of access, existance of bounty etc. CargoBountyConsoleState now requires a new parameter - untilNextSkip, which is used to disable the button and display the cooldown clientside.

## Media

https://youtu.be/XGMzkcljD5M
image
image

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- add: Quartermasters can now skip a single bounty in the list once every 15 minutes.

### CODE REVIEW

**Comment by Adrian16199** [comment]
God bless.
No more pain of not being able to find that last soap.

**Comment by Terraspark4941** [comment]
cargonia goes from being german-productive to being japanese-productive ongod

File: Content.Client/Cargo/UI/BountyEntry.xaml.cs [hunk 1]
```diff
@@ -38,6 +39,20 @@ public BountyEntry(CargoBountyData bounty)
         DescriptionLabel.SetMarkup(Loc.GetString("bounty-console-description-label", ("description", Loc.GetString(bountyPrototype.Description))));
         IdLabel.SetMarkup(Loc.GetString("bounty-console-id-label", ("id", bounty.Id)));
 
-        PrintButton.OnPressed += _ => OnButtonPressed?.Invoke();
+        PrintButton.OnPressed += _ => OnLabelButtonPressed?.Invoke();
+        SkipButton.OnPressed += _ => OnSkipButtonPressed?.Invoke();
+    }
+
+    public void UpdateSkipButton(TimeSpan untilNextSkip)
```
Comment (EmoGarbage404) [inline Content.Client/Cargo/UI/BountyEntry.xaml.cs:46]: you can make this update in real time if you
- store `untilNextSkip` inside of this class
- just call this function via `FrameUpdate()`

that way you can watch the timer tick down.

File: Content.Server/Cargo/Components/StationCargoBountyDatabaseComponent.cs [hunk 2]
```diff
@@ -32,4 +33,16 @@ public sealed partial class StationCargoBountyDatabaseComponent : Component
     /// </summary>
     [DataField]
     public HashSet<string> CheckedBounties = new();
+
+    /// <summary>
+    /// The time at which players will be able to skip the next bounty.
+    /// </summary>
+    [DataField("nextSkipTime", customTypeSerializer: typeof(TimeOffsetSerializer))]
```
Comment (EmoGarbage404) [inline Content.Server/Cargo/Components/StationCargoBountyDatabaseComponent.cs:40]:
```suggestion
    [DataField(customTypeSerializer: typeof(TimeOffsetSerializer))]
```

File: Content.Server/Cargo/Components/StationCargoBountyDatabaseComponent.cs [hunk 3]
```diff
@@ -32,4 +33,16 @@ public sealed partial class StationCargoBountyDatabaseComponent : Component
...
+    public TimeSpan NextSkipTime = TimeSpan.Zero;
+
+    /// <summary>
+    /// The time between skipping bounties.
+    /// </summary>
+    [DataField("skipDelay")]
```
Comment (EmoGarbage404) [inline Content.Server/Cargo/Components/StationCargoBountyDatabaseComponent.cs:46]:
```suggestion
    [DataField]
```

**Review by EmoGarbage404 — CHANGES_REQUESTED** [review]

File: Content.Server/Cargo/Components/StationCargoBountyDatabaseComponent.cs (hunk 3 above)
Comment (Killerqu00) [reply Content.Server/Cargo/Components/StationCargoBountyDatabaseComponent.cs:46]: ConsoleComponent has an argument for DataField, but DatabaseComponent should not? Is there a rule for when I should and shouldn't do that?

Comment (deltanedas) [reply Content.Server/Cargo/Components/StationCargoBountyDatabaseComponent.cs]: the tag name is now autogenerated from the name
so SkipDelay first letter lowercase is skipDelay, so you dont need it

Comment (Killerqu00) [reply Content.Server/Cargo/Components/StationCargoBountyDatabaseComponent.cs]: ahh got it, thanks

**Review by EmoGarbage404 — APPROVED** [review]

**Pull request merged**